messaging/src/main/java/org/axonframework/messaging/FluxMessageStream.java

```diff
@@ -66,6 +80,89 @@ public <R> CompletableFuture<R> reduce(@Nonnull R identity,
         return source.reduce(identity, accumulator).toFuture();
     }
 
+    @Override
+    public Optional<Entry<M>> next() {
+        subscribeToSource();
+        Entry<M> poll = readAhead.poll();
+        if (poll != null && readAhead.isEmpty()) {
+            if (!cancelled.get()) {
+                subscription.get().request(1);
+            }
+        }
+        return Optional.ofNullable(poll);
+    }
+
+    @Override
+    public void onAvailable(@Nonnull Runnable callback) {
+        this.availabilityCallback.set(callback);
+        if (hasNextAvailable() || isCompleted()) {
+            callback.run();
+        }
+        subscribeToSource();
+    }
+
+    @Override
+    public Optional<Throwable> error() {
+        if (readAhead.isEmpty()) {
+            return Optional.ofNullable(error.get());
+        }
+        // there is still data to read, so we're not reporting an error
+        return Optional.empty();
+    }
+
+    @Override
+    public boolean isCompleted() {
+        return readAhead.isEmpty() && completed.get();
+    }
+
+    @Override
+    public boolean hasNextAvailable() {
```

Comment:
Doesn't `hasNextAvailable` also require us to subscribe to the source with `subscribeToSource()`?

Reply:
Yes, correct. Well spotted.